Improve the GraphQL and gRPC Get Started Guides

praneesha opened the pull request:
## Purpose
Review the GraphQL and gRPC Get Started guides.
> Fixes #

## Checklist

- [ ] **Page addition**
  - [ ] Add `permalink` to pages.

- [ ] **Page removal**
  - [ ] Remove entry from corresponding left nav YAML file.
  - [ ] Add `redirect_from` on the alternative page.
  - [ ] If no alternative page, add redirection on the `redirections.js ` file.

- [ ] **Page rename**
  - [ ] Add front-matter `redirect_from`.
  - [ ] Add front-matter `redirect_to:` (if applicable).

- [ ] **Page restrcuture**
  - [ ] Add `permalink` to pages.
  - [ ] Add front-matter `redirect_from`.
  - [ ] Add front-matter `redirect_to:` (if applicable).

anupama-pathirage commented:
@praneesha Seems this PR has some commits related to grpc as well. Please check.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
-GraphQL service (schema-first approach) while others do not need the schema to create the service
-(code-first approach). Ballerina GraphQL package uses the latter. Therefore, you do not need the schema file to create
-your service, and instead, once you write the Ballerina service, the Ballerina GraphQL package will generate the schema.
+Usually, a GraphQL endpoint is defined using a GraphQL schema. Even though some languages require the GraphQL schema to create a GraphQL service (schema-first approach), Ballerina do not need the schema to create the service
```
```suggestion
Usually, a GraphQL endpoint is defined using a GraphQL schema. Even though some languages require the GraphQL schema to create a GraphQL service (schema-first approach), Ballerina does not need the schema to create the service
```

praneesha replied:
Fixed.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
-The GraphQL endpoint you are going to create in this guide will have two main operations, `Query` and `Mutation`. The
-`Query` type will be used to read the data in the data source, and the `Mutation` operation will be used to update the
-data in the data source.
+The GraphQL endpoint you are going to create in this guide will have two main operations as `Query` and `Mutation`. The `Query` type will be used to read the data in the data source, and the `Mutation` operation will be used to update the data in the data source.
```
```suggestion
The GraphQL endpoint you will create in this guide will have two main operations: `Query` and `Mutation`. The `Query` type will be used to read the data in the data source, and the `Mutation` operation will be used to update the data in the data source.
```

anupama-pathirage commented on _data/graphapiguidenavswanlake.yml:
```diff
@@ -4,15 +4,33 @@
+      active: try-the-service
+    - title: Learn more
+      url: /learn/write-a-graphql-api-with-ballerina/#learn-more
+      active: learn-more
```
Missign new line

praneesha replied:
Fixed.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -143,21 +162,29 @@ public distinct service class CovidData {
 cannot be `null` in the GraphQL response. In other words, these fields have `NON_NULL` types. (In GraphQL, these are
-represented by the exclamation mark `!`. E.g., `String!`). However, the resource methods returning numbers can return `null`
-values. Therefore, the type of fields represented by those resource methods is nullable.
+represented by the exclamation mark `!`. E.g., `String!`). However, the resource methods returning numbers can return `null`values. Therefore, the type of fields represented by those resource methods is nullable.
```
```suggestion
represented by the exclamation mark `!`. E.g., `String!`). However, the resource methods returning numbers can return `null` values. Therefore, the type of fields represented by those resource methods is nullable.
```

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -338,15 +375,19 @@ service /covid19 on new graphql:Listener(9000) {
-Now, you can run this service to serve a GraphQL API to our data source. To run, execute the following command:
+## Run the service
+
+ Execute the command below to run this service to serve a GraphQL API to the data source.
```
Output of the run command is missing.

praneesha replied:
Added.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -416,3 +459,11 @@ The result of this request is the following JSON.
+To learn more about GraphQL services in Ballerina, see the following.
+
+- [`graphql` library documentation](https://lib.ballerina.io/ballerina/graphql/latest)
+- [Hello World](/learn/by-example/graphql-hello-world.html)
```
```suggestion
- [GraphQL Hello World](/learn/by-example/graphql-hello-world.html)
```

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -416,3 +459,11 @@ The result of this request is the following JSON.
+
+- [`graphql` library documentation](https://lib.ballerina.io/ballerina/graphql/latest)
+- [Hello World](/learn/by-example/graphql-hello-world.html)
+- [GraphQL to Ballerina](/learn/cli-documentation/graphql/#graphql-to-ballerina)
```
Shouldn't we add https://ballerina.io/learn/graphql-client-tool/ too?

praneesha replied:
Added.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -416,3 +459,11 @@ The result of this request is the following JSON.
+- [`graphql` library documentation](https://lib.ballerina.io/ballerina/graphql/latest)
+- [Hello World](/learn/by-example/graphql-hello-world.html)
+- [GraphQL to Ballerina](/learn/cli-documentation/graphql/#graphql-to-ballerina)
+
```
Extra new line.

praneesha replied:
Fixed.

praneesha commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -379,19 +420,21 @@ type Mutation {
+- The request body contains the GraphQL query.
 
-The result of this request is the following JSON.
+You view the the JSON below as the result of this request.
```
Fixed.

praneesha commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -192,15 +221,19 @@ service /covid19 on new graphql:Listener(httpListener) {
+>**Info:** This is as same as the first code snippet above, which will listen on the port `9000` and serve on `/covid19`.
 
-Now, create the service methods.
+### Create the service methods
```
@anupama-pathirage - This was added to being consistent with the other guides. Should this still be changed?

anupama-pathirage replied:
It is not exactly creating but implementing. Also, I see it aligns with the previously refactored HTTP guide as well. https://ballerina.io/learn/write-a-restful-api-with-ballerina/#implement-the-first-endpoint . I don't see an inconsistency here.

praneesha commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -416,3 +459,11 @@ The result of this request is the following JSON.
+To learn more about GraphQL services in Ballerina, see the following.
+
+- [`graphql` library documentation](https://lib.ballerina.io/ballerina/graphql/latest)
+- [Hello World](/learn/by-example/graphql-hello-world.html)
```
Changed to `module`.

anupama-pathirage commented on _data/graphapiguidenavswanlake.yml:
```diff
@@ -4,15 +4,34 @@
+    - title: Learn more
+      url: /learn/write-a-graphql-api-with-ballerina/#learn-more
+      active: learn-more
+           
```
Seems incorrect new lines.

praneesha replied:
Fixed.

anupama-pathirage commented:
> > @praneesha Seems this PR has some commits related to grpc as well. Please check.
> 
> Yes, that was intentional. Added 2 minor ones only.

Then PR title is incorrect :)

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
-The `Query` type has two fields.
+### The `query` type
+The `query` type has two fields.
+* Can be identified by the `resource` keyword in the function description.
```
Is it correct to say `function description` ?

praneesha replied:
Removed the sentence below.

`Can be identified by the resource keyword in the function description.`

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
````diff
@@ -14,49 +14,72 @@ redirect_from:
+Created new package 'graphql' at graphql.
+```
+
+This creates a directory named `graphql` with the default module along with a sample code for the service as shown below.
````
`graphql` is not a good name for the package.

praneesha replied:
Updated the package name to `covid19`.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
- `CovidEntry`. A single entry will include the ISO code for the location (`isoCode`), name of the country (`country`),
- number of cases (`cases`), number of deaths (`deaths`), number of recovered cases (`recovered`), and the number of
- active cases (`active`).
+To define the data type to use in the table, replace the API template file (i.e., `main.bal`) with the code below. 
```
in the previous tip it was asekd to remove the `main.bal`. Also if we use `-t service` as I suggested in a previous comment there won't be `main.bal` file.

praneesha replied:
Updated all the occurrences of the `main.bal` file to `service.bal`.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -79,23 +107,14 @@ table<CovidEntry> key(isoCode) covidEntriesTable = table [
-In this guide, the endpoint is going to return the number of cases in thousands. Therefore, a service type is used to
-define the output object type, and inside the service type, each resource function will return the original value
-divided by 1000.
+Since the data map to a GraphQL `Object`, you need to define this object type. In Ballerina, a GraphQL output object can be defined using either a service type or a record type.
```
The existing text `Since the data used in this guide map to a GraphQL Object, you first have to define this object type.` makes more sense to me.

praneesha replied:
Updated.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -143,21 +162,29 @@ public distinct service class CovidData {
-the data set. Each resource method in this service represents a field of the GraphQL object type.  The return type of
+In this code:
+- The endpoint returns the number of cases in thousands. Therefore, a service type is used to define the output object type, and inside the service type, each resource function will return the original value
+divided by 1000.
```
```suggestion
divided by `1000`.
```

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -143,21 +162,29 @@ public distinct service class CovidData {
+In this code:
+- The endpoint returns the number of cases in thousands. Therefore, a service type is used to define the output object type, and inside the service type, each resource function will return the original value
+divided by 1000.
+- You are creating the `CovidData` service type to represent the GraphQL `Object` type, which represents an entry in
```
Why only this point is started as `You are`?

praneesha replied:
Changed to `The `CovidData` service type represents the GraphQL `Object` type, which represents an entry in
the data set. `

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -143,21 +162,29 @@ public distinct service class CovidData {
+
+#### Import the Ballerina GraphQL package
+
+To import the Ballerina GraphQL package for creating the listener object, add the code below to the API template file (i.e., `main.bal`). 
```
I think the titles are incorrect. The code snippet is not for importing the package.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -167,19 +194,21 @@ service /covid19 on new graphql:Listener(9000) {
-In the above code snippet, the path of this service is defined as `/covid19`. If you want to host the service on the
-root, you can remove the path as the following code snippet:
+In this code:
+- The path of this service is defined as `/covid19`. If you want to host the service on the
```
Why adding a bullet here?

praneesha replied:
It was added to be consistent with the other `In this code:` sections. @anupama-pathirage

praneesha replied:
Removed the bullet point.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
````diff
@@ -167,19 +194,21 @@ service /covid19 on new graphql:Listener(9000) {
+    }
+    ```
+
+#### Define the port
````
Incorrect title. This code block and section is not about defining port, but about how to use exiting listneer.See more comments below for this section,

praneesha replied:
Removed.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -192,15 +221,19 @@ service /covid19 on new graphql:Listener(httpListener) {
+
+##### Create the `all` field
+
+To create the `all` field, which returns an array of `CovidData` type, add the code below to the API template file (i.e., `main.bal`). 
```
API template is not appropriate IMO

ThisaruGuruge replied:
Agreed. API template is something different.

praneesha replied:
Removed the word `API template` and changed it to `the `service.bal` file`.

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
````diff
@@ -338,15 +375,37 @@ service /covid19 on new graphql:Listener(9000) {
+```bash
+Compiling source
+	praneesha/graphql:0.1.0
+WARNING [main.bal:(25:5,25:5)] concurrent calls will not be made to this method since the service and the method are not 'isolated'
````
Can we improve the code not to have this warning? @shafreenAnfar , @ThisaruGuruge

ThisaruGuruge replied:
We can, but to do this, we have to add the `isolated` keyword. We decided not to add this since it might be confusing for new users. @shafreenAnfar WDYT?

shafreenAnfar replied:
Yeah, adding the word `isolated` is problematic. But we might be able to do it without it, any thoughts? @MaryamZi

anupama-pathirage commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -338,15 +375,37 @@ service /covid19 on new graphql:Listener(9000) {
+
+## View the generated schema
+
+The console has warning logs related to the isolatedness of resources. It is a built-in service concurrency safety feature of Ballerina.
```
This sentence is not for this section, it talks about the previous section.

praneesha replied:
Fixed.

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
-The `Mutation` type has a single field.
+### The `mutation` type
+The `mutation` type has a single field.
+* Can be identified by the `remote` keyword in the function description.
```
Same as above.

praneesha replied:
Removed.

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
 
-### The `Query` Type
-The `Query` type has two fields.
+### The `query` type
```
The `Query` type is a pre-defined type name, it should start with a capital letter.

```suggestion
### The `Query` type
```

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
 
-### The `Mutation` Type
-The `Mutation` type has a single field.
+### The `mutation` type
```
```suggestion
### The `Mutation` type
```

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
````diff
@@ -14,49 +14,72 @@ redirect_from:
+In the terminal, execute the command below to create the Ballerina package for the API implementation.
+
+```bash
+$ bal new graphql
````
The package name `graphql` is not correct IMO. A better name would be `covid19`.

praneesha replied:
Changed to `$ bal new covid19 -t service`.

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -143,21 +162,29 @@ public distinct service class CovidData {
 
-Now you are all set to write the GraphQL service. To write the service, you need to create a GraphQL listener object.
-To create a listener object, you need to import the Ballerina GraphQL package.
+### Create the listener object
```
Yes, the whole section is re-arranged incorrectly.

praneesha replied:
Fixed.

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -192,15 +221,19 @@ service /covid19 on new graphql:Listener(httpListener) {
 
 As per the design, there are two fields in the `Query` type and one field in the `Mutation` type in your GraphQL
-service. The fields of the `Query` type are represented by resource methods in Ballerina, while The fields of the
+service. The fields of the `Query` type are represented by resource methods in Ballerina, while the fields of the
 `Mutation` type are represented by the remote methods in Ballerina.
```
```suggestion
service. The fields of the `Query` type are represented by the resource methods with the `get` accessor in Ballerina, while the fields of the `Mutation` type are represented by the remote methods in Ballerina.
```

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -213,18 +246,16 @@ service /covid19 on new graphql:Listener(9000) {
-`Query` type. No other accessor is allowed. Then comes the name of the field. The return type is the type of the field.
+In this code:
+- The resource method definition has the accessor `get`, which is used to identify the resource methods as a field of the`query` type. 
+- No other accessor is allowed. Then comes the name of the field. The return type is the type of the field.
```
We now do allow other accessors. (`subscribe`).

```suggestion
- Then comes the name of the field. The return type is the type of the field.
```

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -236,13 +267,15 @@ resource function get filter(string isoCode) returns CovidData? {
+- The `filter` field is defined in the root `query` type. Since this field has an input parameter `isoCode`, you have to add an input parameter to the resource method. 
+- This method returns the corresponding data for the given `isoCode` if such data is available in the data set, and it returns `null` otherwise.
+
+#### Create mutation type remote method 
```
`mutation type remote method` is not accurate. The remote method is used to define a field inside the `Mutation` type.

praneesha replied:
Updated to `Create `Mutation` type remote function`.

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -236,13 +267,15 @@ resource function get filter(string isoCode) returns CovidData? {
+#### Create mutation type remote method 
 
-As the `Query` type is completed now, you need to define the `Mutation` type using remote methods.
+As the `query` type is completed now, define the `mutation` type using remote methods.
```
```suggestion
As the `Query` type is completed now, define the `Mutation` type using remote methods.
```

praneesha commented:
> > > @praneesha Seems this PR has some commits related to grpc as well. Please check.
> > 
> > 
> > Yes, that was intentional. Added 2 minor ones only.
> 
> Then PR title is incorrect :)

Updated the PR title.

praneesha commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
- `CovidEntry`. A single entry will include the ISO code for the location (`isoCode`), name of the country (`country`),
- number of cases (`cases`), number of deaths (`deaths`), number of recovered cases (`recovered`), and the number of
- active cases (`active`).
+To define the data type to use in the table, replace the API template file (i.e., `main.bal`) with the code below. 
```
Removed the word `API template` and changed it to `the `service.bal` file`.

praneesha commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
-GraphQL service (schema-first approach) while others do not need the schema to create the service
-(code-first approach). Ballerina GraphQL package uses the latter. Therefore, you do not need the schema file to create
-your service, and instead, once you write the Ballerina service, the Ballerina GraphQL package will generate the schema.
+Usually, a GraphQL endpoint is defined using a GraphQL schema. Even though some languages require the GraphQL schema to create a GraphQL service (schema-first approach), Ballerina does not need the schema to create the service
+(code-first approach). Once you write the Ballerina service, the Ballerina GraphQL package will generate the schema.
```
Updated.

praneesha commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -14,49 +14,72 @@ redirect_from:
+3. A command terminal
 
-## Design the GraphQL endpoint
+## Understand the implementation
```
Reverted to `Design the GraphQL endpoint`.

praneesha commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -167,19 +194,21 @@ service /covid19 on new graphql:Listener(9000) {
 
-When creating the `graphql:Listener` object, you need to provide the port to which it is listening. Alternatively, an
-existing `http:Listener` object can also be used to initialize the `graphql:Listener` as below.
+When creating the `graphql:Listener` object, you need to provide the port to which it is listening. Alternatively, to use an existing `http:Listener` object, for initializing the `graphql:Listener`, add the code below to the API template file (i.e., `main.bal`). 
```
Fixed.

ThisaruGuruge commented on _data/graphapiguidenavswanlake.yml:
```diff
@@ -7,12 +7,31 @@
+    - title: Learn more
+      url: /learn/write-a-graphql-api-with-ballerina/#learn-more
+      active: learn-more
+       
```
```suggestion

```

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -192,39 +205,36 @@ service /covid19 on new graphql:Listener(httpListener) {
-The resource method definition has the accessor `get`, which is used to identify the resource methods as a field of the
-`Query` type. No other accessor is allowed. Then comes the name of the field. The return type is the type of the field.
+In this code:
+- The resource method definition has the accessor `get`, which is used to identify the resource methods as a field of the`query` type. 
```
```suggestion
- The resource method definition has the accessor `get`, which is used to identify the resource methods as a field of the `Query` type. 
```

ThisaruGuruge commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -236,13 +246,17 @@ resource function get filter(string isoCode) returns CovidData? {
-input parameter `isoCode`, you have to add an input parameter to the resource method. This method returns the
-corresponding data for the given `isoCode` if such data is available in our data set, otherwise, it returns `null`.
+In this code:
+- The `filter` field is defined in the root `query` type. Since this field has an input parameter `isoCode`, you have to add an input parameter to the resource method. 
```
Please check all the occurrences. The `Query` is a predefined type that should start with a capital letter.

```suggestion
- The `filter` field is defined in the root `Query` type. Since this field has an input parameter `isoCode`, you have to add an input parameter to the resource method. 
```

praneesha commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -192,39 +205,36 @@ service /covid19 on new graphql:Listener(httpListener) {
+As per the design, there are two fields in the `Query` type and one field in the `Mutation` type in your GraphQL service. The fields of the `Query` type are represented by the resource methods with the `get` accessor in Ballerina, while the fields of the `Mutation` type are represented by the remote methods in Ballerina.
 
-Let's first create the `all` field. This should return an array of `CovidData` type.
+#### Create `Query` type resource functions
```
Fixed.

praneesha commented on learn/get-started/write-a-graphql-api-with-ballerina.md:
```diff
@@ -236,13 +246,17 @@ resource function get filter(string isoCode) returns CovidData? {
+- The `filter` field is defined in the root `Query` type. Since this field has an input parameter `isoCode`, you have to add an input parameter to the resource method. 
+- This method returns the corresponding data for the given `isoCode` if such data is available in the data set, and it returns `null` otherwise.
+
+#### Create `Mutation` type remote function
```
Fixed.

ThisaruGuruge reviewed (approved):
Reviewed the GraphQL guide. LGTM.
